src/celeritas/optical/CerenkovParams.cc

```diff
@@ -0,0 +1,62 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/CerenkovParams.cc
+//---------------------------------------------------------------------------//
+#include "CerenkovParams.hh"
+
+#include <utility>
+#include <vector>
+
+#include "corecel/cont/Range.hh"
+#include "corecel/data/CollectionBuilder.hh"
+#include "corecel/math/Algorithms.hh"
+#include "celeritas/Quantities.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/grid/GenericGridData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with optical property data.
+ */
+CerenkovParams::CerenkovParams(HostCRef<OpticalPropertyData> const& properties)
+{
+    HostVal<CerenkovData> data;
+    for (auto mat_id :
+         range(OpticalMaterialId(properties.refractive_index.size())))
+    {
+        GenericGridData ai_grid;
+        auto const& ri_grid = properties.refractive_index[mat_id];
+        if (!ri_grid)
+        {
+            // No refractive index data stored for this material
+            make_builder(&data.angle_integral).push_back(ai_grid);
+            continue;
+        }
+
+        // Calculate the Cerenkov angle integral
+        auto const refractive_index = properties.reals[ri_grid.value];
+        auto const energy = properties.reals[ri_grid.grid];
+        std::vector<real_type> integral(energy.size());
+        for (size_type i = 1; i < energy.size(); ++i)
+        {
+            integral[i] = integral[i - 1]
+                          + real_type(0.5) * (energy[i] - energy[i - 1])
+                                * (1 / ipow<2>(refractive_index[i - 1])
+                                   + 1 / ipow<2>(refractive_index[i]));
+        }
+        auto reals = make_builder(&data.reals);
```

Comment:
... and then using a `DedupeCollectionBuilder reals` at the parent scope will allow the same data to be automatically reused if the grids or refraction indices are identical.